internal/pkg/kubernetes/client.go

```diff
@@ -0,0 +1,360 @@
+	podLogOpts := v1.PodLogOptions{
+		Follow: true,
+	}
+
+	ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	request := client.clientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &podLogOpts)
+	response, err := request.Stream(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func (client *kubeClient) WaitForReadyJob(executionName string, waitTime time.Duration) error {
+	jobs := client.clientSet.BatchV1().Jobs(client.namespace)
+	listOptions := meta.ListOptions{
+		TypeMeta:      typeMeta,
+		LabelSelector: jobLabelSelector(executionName),
+	}
+
+	ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	var err error
+	for i := 0; i < client.kubeWaitForResourcePollCount; i += 1 {
+		watchJob, watchErr := jobs.Watch(ctx, listOptions)
+		if watchErr != nil {
+			err = watchErr
+			continue
+		}
+
+		timeoutChan := time.After(waitTime)
+		resultChan := watchJob.ResultChan()
+
+		var job *batch.Job
+		for {
+			select {
+			case event := <-resultChan:
+				if event.Type == watch.Error {
+					err = watcherError("jobs", listOptions)
+					break
+				}
+
+				// Ignore empty events
+				if event.Object == nil {
+					continue
+				}
+
+				job = event.Object.(*batch.Job)
+				if job.Status.Active >= 1 || job.Status.Succeeded >= 1 || job.Status.Failed >= 1 {
+					watchJob.Stop()
+					return nil
+				}
+			case <-timeoutChan:
+				err = timeoutError
+				break
+			}
+			if err != nil {
+				watchJob.Stop()
+				break
+			}
+		}
+	}
+
+	return err
+}
+
+func watcherError(resource string, listOptions meta.ListOptions) error {
+	return fmt.Errorf("watch error when waiting for %s with list option %v", resource, listOptions)
+}
+
+func (client *kubeClient) WaitForReadyPod(executionName string, waitTime time.Duration) (*v1.Pod, error) {
+	coreV1 := client.clientSet.CoreV1()
+	kubernetesPods := coreV1.Pods(client.namespace)
+	listOptions := meta.ListOptions{
+		LabelSelector: jobLabelSelector(executionName),
+	}
+
+	var err error
+	ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	for i := 0; i < client.kubeWaitForResourcePollCount; i += 1 {
+		watchJob, watchErr := kubernetesPods.Watch(ctx, listOptions)
+		if watchErr != nil {
+			err = watchErr
+			continue
+		}
+
+		timeoutChan := time.After(waitTime)
+		resultChan := watchJob.ResultChan()
+
+		var pod *v1.Pod
+		for {
+			select {
+			case event := <-resultChan:
+				if event.Type == watch.Error {
+					err = watcherError("pods", listOptions)
+					watchJob.Stop()
+					break
+				}
+
+				// Ignore empty events
+				if event.Object == nil {
+					continue
+				}
+
+				pod = event.Object.(*v1.Pod)
+				if pod.Status.Phase == v1.PodRunning || pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed {
+					watchJob.Stop()
+					return pod, nil
+				}
+			case <-timeoutChan:
+				err = timeoutError
+				watchJob.Stop()
+				break
+			}
+			if err != nil {
```

Comment:
why there is error handling here?

why not immediately handle the error on `case <-timeoutChan` ?

Reply:
so this tracks the error we get on the watch channel. Inside the timeout we track for timeout of request.